check volume directories instead of mounts for cleanupOrphanedPodDirs

mucahitkurt opened the pull request:
**What type of PR is this?**
/kind cleanup
/sig storage

**What this PR does / why we need it**:
Related comment; https://github.com/kubernetes/kubernetes/issues/72346#issuecomment-467330114

Change `podVolumesExist` to check volume directories instead of mounted volumes when it's called from `Kubelet.cleanupOrphanedPodDirs`. Because `getMountedVolumePathListFromDisk` does not always produce correct results for mounted volumes and `Kubelet.cleanupOrphanedPodDirs` already used `getPodVolumePathListFromDisk`.

**Which issue(s) this PR fixes**:
Related issue https://github.com/kubernetes/kubernetes/issues/72346

Related issue is quite evolved to fix all `podVolumesExist` method usage, but we may think this PR a fix for the issue https://github.com/kubernetes/kubernetes/issues/72346 and all `podVolumesExist` usage may be fixed within the scope of issue https://github.com/kubernetes/kubernetes/issues/72347

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```
cc @msau42 @cofyc

mucahitkurt committed:
Change the logic of pod volumes existence check during kubelet cleanupOrphanedPodDirs, cleanupOrphanedPodCgroups and PodResourcesAreReclaimed

check in-memory cache whether volumes are still mounted and check disk directory for the volume paths instead of mounted volumes check

Signed-off-by: Mucahit Kurt <mucahitkurt@gmail.com>

msau42 commented:
/assign @msau42 @jsafrane

mucahitkurt commented:
I'm looking into build error, when I run `./hack/update-bazel.sh` it doesn't update any build file...

mattjmcnaughton commented on pkg/kubelet/kubelet_pods.go:
```diff
@@ -917,7 +917,7 @@ func (kl *Kubelet) PodResourcesAreReclaimed(pod *v1.Pod, status v1.PodStatus) bo
 		return false
 	}
-	if kl.podVolumesExist(pod.UID) && !kl.keepTerminatedPodVolumes {
+	if kl.podVolumesExist(pod.UID, false) && !kl.keepTerminatedPodVolumes {
```
Nit: Can we do
```
checkVolumePaths := false
kl.podVolumesExist(pod.UID, checkVolumePaths)
```

otherwise, it can be difficult to know what the `false` means here.

mucahitkurt replied:
I added parameter name as a comment for `podVolumesExist` call. Thanks!

mattjmcnaughton replied:
Thanks for working with me on this small nit :)

This may just be my opinion, but I feel like the comment offers less advantages than the variable name (i.e. is less explicit). Are there advantages to using a comment over a variable that I'm missing?

mucahitkurt replied:
My IDE(goland) uses similar approach to show the parameter names of the function variables and I see similar usages in kubernetes code base, using an additional variable seems little noisy to me, so I choose adding comment.

But I agree, using variable can be more explicit, I'm neutral, I'll add additional variable instead of comment.

mucahitkurt replied:
Fixed

mattjmcnaughton replied:
Ah gotcha - maybe part of this is that I'm using vim haha :) Anyway, thanks!+

mattjmcnaughton replied:
Will let @jsafrane or someone else on sig-storage sign off :)

mattjmcnaughton reviewed (commented):
A small readability comment from me - otherwise, defer to @jsafrane and the folks who work with volumes more :)

mucahitkurt commented:
/test pull-kubernetes-node-e2e-containerd
/test pull-kubernetes-kubemark-e2e-gce-big

cofyc commented:
/cc @cofyc

mucahitkurt commented:
/test pull-kubernetes-verify

jsafrane commented:
I am not sure I understand the PR. If we expect that all volume plugins remove their directories in their `TearDown` (that's what `cleanupOrphanedPodDirs` assumes now), why we are checking *mounts* in all cleanupOrphanedPodDirs(...,  false)` calls? Shouldn't be the check the same everywhere?

mucahitkurt commented:
> I am not sure I understand the PR. If we expect that all volume plugins remove their directories in their `TearDown` (that's what `cleanupOrphanedPodDirs` assumes now), why we are checking _mounts_ in all cleanupOrphanedPodDirs(..., false)` calls? Shouldn't be the check the same everywhere?

Other callers of `podExists` method, `cleanupOrphanedPodCgroups`, `PodResourcesAreReclaimed`, assume that all plugins _unmount_ their directories but don't assume that all plugins remove their directories. So we are not sure to change this behavior because it can cause not deleting a pod.

But `cleanupOrphanedPodDirs` was checking both _mounts_(via `podVolumeExist` call) and _directory existence_(via `getPodVolumePathListFromDisk`) before this PR, but now it's only checking _directory existence_, so I removed a redundant check and the behavior of the method doesn't change.

jsafrane commented:
> `cleanupOrphanedPodCgroups`,`PodResourcesAreReclaimed`, assume that all plugins unmount their directories but don't assume that all plugins remove their directories.

Now we assume (in `cleanupOrphanedPodDirs`), that volume plugins do remove their directories. 

I see you want be on the safe side and it's probably correct at this stage of 1.17, but should we have the same assumption everywhere in the end?

/lgtm
/approve

mucahitkurt commented:
> Now we assume (in `cleanupOrphanedPodDirs`), that volume plugins do remove their directories.
> 
> I see you want be on the safe side and it's probably correct at this stage of 1.17, but should we have the same assumption everywhere in the end?
> 
> /lgtm
> /approve

I think so, this's [first suggestion of @cofyc](https://github.com/kubernetes/kubernetes/issues/72346#issuecomment-466347067). I think we are not confident about to depending the cleanup behavior of plugins in terms of removing directories but currently these methods already depend the cleanup behavior of plugins in terms of unmounting directories. May be @msau42 or @cofyc can express better than me.

mucahitkurt commented:
/assign @vishh

mucahitkurt commented:
@kubernetes/sig-node-pr-reviews PTAL

mucahitkurt commented:
/test pull-kubernetes-e2e-gce

jsafrane commented:
/lgtm

msau42 commented:
/assign @jingxu97

mucahitkurt commented:
/remove-lifecycle rotten

jingxu97 commented on pkg/kubelet/kubelet_volumes.go:
```diff
@@ -51,12 +51,24 @@ func (kl *Kubelet) ListVolumesForPod(podUID types.UID) (map[string]volume.Volume
 // podVolumesExist checks with the volume manager and returns true any of the
 // pods for the specified volume are mounted.
-func (kl *Kubelet) podVolumesExist(podUID types.UID) bool {
+func (kl *Kubelet) podVolumesExist(podUID types.UID, checkVolumePaths bool) bool {
```
I think it might be more clear to  create another function for this purpose instead of reuse this one and adding a parameter of it. They are calling different functions. kl.getPodVolumePathListFromDisk , kl.getMountedVolumePathListFromDisk

mucahitkurt replied:
I added another function for the `podVolumePathsExist`.

msau42 commented on pkg/kubelet/kubelet_pods.go:
```diff
@@ -918,7 +918,8 @@ func (kl *Kubelet) PodResourcesAreReclaimed(pod *v1.Pod, status v1.PodStatus) bo
 		return false
 	}
-	if kl.podVolumesExist(pod.UID) && !kl.keepTerminatedPodVolumes {
+	checkVolumePaths := false
```
@jingxu97 I believe we want to change this check to check for directory existence instead of mount point right?

mucahitkurt replied:
Hey @msau42 , Should I continue to this PR, is it still relevant?

msau42 replied:
Hi @mucahitkurt yes this issue still exists. Fixing this will also fix this flaky test: https://github.com/kubernetes/kubernetes/issues/91001

mucahitkurt replied:
I added `podVolumePathsExist` check here.

dashpole commented:
/assign

mucahitkurt commented:
/test pull-kubernetes-e2e-kind

mucahitkurt commented:
@jingxu97 , @msau42 PTAL, thanks!

jingxu97 commented:
@mucahitkurt sorry for the delay. Could you rebase your code? I will review it now.

jingxu97 commented on pkg/kubelet/kubelet_volumes.go:
```diff
@@ -73,6 +73,27 @@ func (kl *Kubelet) podVolumesExist(podUID types.UID) bool {
+// volumes for the specified pod are mounted or any of the volume paths of the specified pod exist.
+func (kl *Kubelet) podVolumePathsExist(podUID types.UID) bool {
+	if mountedVolumes :=
+		kl.volumeManager.GetMountedVolumesForPod(
```
this is also checked in podVolumesExist() function above?

mucahitkurt replied:
Yes, I didn't want to create a coupling between these two methods `podVolumesExist` and `podVolumePathsExist` even though all current use cases of `podVolumePathsExist` also use the `podVolumesExist`.

Should I remove this check from `podVolumePathsExist`?

mucahitkurt commented:
> @mucahitkurt sorry for the delay. Could you pls rebase your code? I will review it now.

Hey @jingxu97 I have rebased this PR, PTAL.

mucahitkurt commented:
/test pull-kubernetes-bazel-test

jingxu97 commented:
@mucahitkurt sorry for the delay. I think the main logic here is changing the ordering of checking things. There are three places we use to check pod volume still exist

1. kl.volumeManager.GetMountedVolumesForPod(podUid)
   This will check in-memory cache whether volumes are still mounted. This check is fast, but might not reliable when kubelet restarts and lost its cache (there is a reconstruction phase for cache from scanning dirs on disk, but I need to check whether there is guarantee when cleanupOrphanedPodDirs is called, cache is already constructed)

2. tbd

mucahitkurt commented:
@jingxu97, @msau42 Sorry for the delay, I'll try to check the latest comments and try to make the fix in a few days (hope before 12th Nov.)

mucahitkurt commented:
> @mucahitkurt sorry for the delay. I think the main logic here is changing the ordering of checking things. There are three types of checking for pod volumes existence, check in-memory cache, check disk directories, and check whether each dir is a mount point or not.
> 
> 1. kl.volumeManager.GetMountedVolumesForPod(podUid)
>    This will check in-memory cache whether volumes are still mounted. This check is fast, but might not be completely reliable when kubelet restarts and lost its cache (there is a reconstruction phase for cache from scanning dirs on disk, but I need to check whether there is guarantee when cleanupOrphanedPodDirs is called, cache is already constructed)
> 2. kl.getPodVolumePathListFromDisk(podUid)
>    This function directly scans the disk directory and returns a list of the volume paths for the given pod (e.g., /var/lib/kubelet/pods/pod_uid/volumes/...)
> 3. kl.getMountedVolumePathListFromDisk(podUID)
>    This function first calls getPodVolumePathListFromDisk to get the list of volume paths for the given pod, and check each path whether is a mount point or not using IsLikelyNotMountPoint.
>    There are two things worth noting about this function
>    
>    * this check might not work in certain cases that mount point is in the root file system).
>    * this check uses os.Stats which might cause process hung in case of NFS mount point when server is down. (Please check this very old and long discussion #31272)
>    * there is a similar function mount.IsNotMountPoint(), this function checks mount table for further verification about mount point to solve the issue with IsLikelyNotMountPoint, but it might take longer time to go through mount table.
> 4. kl.podVolumesExist(podUID)
>    This function is a combination of function 1 and 3 listed here, it first calls function GetMountedVolumesForPod and getMountedVolumePathListFromDisk to verify volume existence.
> 
> All these checking are used in a number of difference places. Here clearupOrphanedPodDirs, this function is being called by HandlePodCleanups, and its comment mentioned
> 
> ```
> // NOTE: This function is executed by the main sync loop, so it
> // should not contain any blocking calls.
> ```
> 
> You might check the issue #31272 to see how we removed the check 3 getMountedVolumePathListFromDisk due to hung issue caused by nfs. However, I think I accidentally introduced this back again here by adding the check of mounted points in podVolumesExist. So the fix here I suggest, instead of calling podVolumesExist, call the following checks
> 
> 1. call check 1 kl.volumeManager.GetMountedVolumesForPod(podUid) for check internal cache,
> 2. call check 2 kl.getPodVolumePathListFromDisk(podUid)
> 
> Pls let me know if you have questions about the comments. Thanks!

@jingxu97 ;

I keep the new function `kl.podVolumePathsExist(podUID types.UID)` because it uses the check 1 and check 2.

I replaced the `kl.podVolumesExist(podUID types.UID)` with `kl.podVolumePathsExist(podUID types.UID)` in `cleanupOrphanedPodDirs` and `PodResourcesAreReclaimed`.

Do you have any concern with the new function?

`kl.podVolumesExist(podUID types.UID)` is used in `Kubelet.cleanupOrphanedPodCgroups`, I keep it as is.

jingxu97 commented:
@mucahitkurt Thanks for the update. The function looks good to me. The only thing is the naming, podVolumePathsExist and podVolumesExist, they are too close so it is a little bit hard to understand what's the difference. I am not good at naming either, just a suggestion, e.g., 

podVolumePathsExist --> podVolumePathsExistInCacheOrDisk
podVolumesExist --> podMountedVolumesExistInCacheOrDisk

mucahitkurt commented:
> @mucahitkurt Thanks for the update. The function looks good to me. The only thing is the naming, podVolumePathsExist and podVolumesExist, they are too close so it is a little bit hard to understand what's the difference. I am not good at naming either, just a suggestion, e.g.,
> 
> podVolumePathsExist --> podVolumePathsExistInCacheOrDisk
> podVolumesExist --> podMountedVolumesExistInCacheOrDisk

@jingxu97 I renamed with your suggestions, I don't have better alternatives.

mucahitkurt commented:
/test pull-kubernetes-e2e-gce-ubuntu-containerd

jingxu97 commented on pkg/kubelet/kubelet_pods.go:
```diff
@@ -1962,7 +1962,7 @@ func (kl *Kubelet) cleanupOrphanedPodCgroups(pcm cm.PodContainerManager, cgroupP
 		// process in the cgroup to the minimum value while we wait.  if the kubelet
 		// is configured to keep terminated volumes, we will delete the cgroup and not block.
-		if podVolumesExist := kl.podVolumesExist(uid); podVolumesExist && !kl.keepTerminatedPodVolumes {
+		if podVolumesExist := kl.podMountedVolumesExistInCacheOrDisk(uid); podVolumesExist && !kl.keepTerminatedPodVolumes {
```
actually here cleanupOrphanedPodCgroups,I think it should be similar to cleanupOrphanedPod, so using podVolumePathsExistInCacheOrDisk. ?

mucahitkurt replied:
I changed this one to `kl. kl.podVolumePathsExistInCacheOrDisk`.

Right now no usage remain for the function `kl.podVolumesExist` or with the new name `kl.podMountedVolumesExistInCacheOrDisk`. Should I delete this function and all related test codes?

mucahitkurt commented:
/test pull-kubernetes-e2e-gce-ubuntu-containerd

jingxu97 commented:
/lgtm

derekwaynecarr commented:
/approve

msau42 commented:
/retest

msau42 commented:
/retest
